libwacom/libwacom-database.c

```diff
@@ -295,32 +412,71 @@ libwacom_parse_stylus_keyfile(WacomDeviceDatabase *db, const char *path)
 
 	groups = g_key_file_get_groups (keyfile, NULL);
 	for (i = 0; groups[i]; i++) {
-		WacomStylus *stylus;
+		WacomStylus *stylus = NULL, *aliased = NULL;
 		WacomStylusId id;
 		GError *error = NULL;
 		char *eraser_type, *type;
 		char **string_list;
+		int aliased_int;
 
 		if (!parse_stylus_id(groups[i], &id)) {
 			g_warning ("Failed to parse stylus ID '%s', ignoring entry", groups[i]);
 			continue;
 		}
+
+		char *aliasstr = g_key_file_get_string(keyfile, groups[i], "AliasOf", NULL);
+
+		if (handle_aliases == IGNORE_ALIASES && aliasstr) {
+			g_clear_pointer(&aliasstr, g_free);
+			continue;
+		}
+
+		if (handle_aliases == ONLY_ALIASES && !aliasstr) {
+			g_clear_pointer(&aliasstr, g_free);
+			continue;
+		}
+
 		stylus = g_new0 (WacomStylus, 1);
 		stylus->refcnt = 1;
 		stylus->id = id;
-		stylus->name = g_key_file_get_string(keyfile, groups[i], "Name", NULL);
-		stylus->group = g_key_file_get_string(keyfile, groups[i], "Group", NULL);
+
+		if (handle_aliases == ONLY_ALIASES) {
+			WacomStylusId lookup;
+
+			if (!parse_stylus_id(groups[i], &lookup)) {
+				g_warning ("Failed to parse Aliased ID '%s', ignoring entry", groups[i]);
+				continue;
+			}
+			if (safe_atoi_base (aliasstr, &aliased_int, 16))
+				lookup.tool_id = aliased_int;
+
+			aliased = g_hash_table_lookup(db->stylus_ht, &lookup);
+			if (!aliased) {
+				g_warning ("Unkown Aliased ID 0x%08x %s\n", lookup.tool_id, aliasstr);
+				continue;
+			}
+		}
+		g_clear_pointer(&aliasstr, g_free);
+
+		stylus->name = string_or_fallback(keyfile, groups[i], "Name", aliased ? aliased->name : NULL, NULL);
+		stylus->group = string_or_fallback(keyfile, groups[i], "Group", aliased ? aliased->group : NULL, NULL);
 		stylus->paired_stylus_ids = g_array_new (FALSE, FALSE, sizeof(WacomStylusId));
 
-		eraser_type = g_key_file_get_string(keyfile, groups[i], "EraserType", NULL);
+		eraser_type = string_or_fallback(keyfile, groups[i], "EraserType", aliased ? eraser_str_from_type(aliased->eraser_type) : NULL, NULL);
 		stylus->eraser_type = eraser_type_from_str (eraser_type);
 		g_clear_pointer(&eraser_type, g_free);
 
 		/* We have to keep the integer array for libwacom_get_supported_styli() */
 		stylus->deprecated_paired_ids = g_array_new (FALSE, FALSE, sizeof(int));
 		stylus->paired_styli = g_array_new (FALSE, FALSE, sizeof(WacomStylus*));
 
-		string_list = g_key_file_get_string_list (keyfile, groups[i], "PairedStylusIds", NULL, NULL);
+		string_list = g_key_file_get_string_list(keyfile, groups[i], "PairedStylusIds", NULL, NULL);
+		if (handle_aliases != IGNORE_ALIASES) {
```

Comment:
```suggestion
		if (handle_aliases == ONLY_ALIASES) {
```
we only have two values, might as well check for the explicit one